### PR TITLE
feat(provider/dcos): Load balancer and secret caching agents

### PR DESCRIPTION
**Pull request by willgorman**
This pull request adds the loadbalancer and secret caching agents for DC/OS along with some related classes and tests.

### CODE REVIEW

File: clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/agent/DcosLoadBalancerCachingAgent.groovy [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    keys = keys.findResults {
+      def parse = Keys.parse(it)
+      if (parse && parse.account == accountName) {
+        return it
+      } else {
+        return null
+      }
+    }
+
+    providerCache.getAll(Keys.Namespace.ON_DEMAND.ns, keys).collect {
```
Comment (duftler) [inline clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/agent/DcosLoadBalancerCachingAgent.groovy]: Please use `return` statement.

File: clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/agent/DcosLoadBalancerCachingAgent.groovy [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+  private List<App> loadLoadBalancers() {
+    // Currently not supporting anything but account global load balancers - no associated region.
+    final Optional<GetAppNamespaceResponse> response = dcosClient.maybeApps(accountName)
+    if (!response.isPresent()) {
+      log.info("The account namespace [${accountName}] does not exist in DC/OS. No load balancers will be cached.")
+      return []
+    }
+
+    response.get().apps.findAll {
```
Comment (duftler) [inline clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/agent/DcosLoadBalancerCachingAgent.groovy]: Please use `return` statement.

File: clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/agent/DcosLoadBalancerCachingAgent.groovy [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+          attributes.name = appId.toString()
+          attributes.app = loadBalancer
+          // Relationships are stored in DcosServerGroupCachingAgent.
+        }
+      }
+    }
+
+    log.info("Caching ${cachedLoadBalancers.size()} load balancers in ${agentType}")
+
+    new DefaultCacheResult([
```
Comment (duftler) [inline clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/agent/DcosLoadBalancerCachingAgent.groovy]: Please use `return` statement.

File: clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/view/DcosLoadBalancerProvider.groovy [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    applicationServerGroups.each { CacheData serverGroup ->
+      loadBalancerKeys.addAll(serverGroup.relationships[Keys.Namespace.LOAD_BALANCERS.ns] ?: [])
+    }
+
+    loadBalancerKeys.addAll(cacheView.filterIdentifiers(Keys.Namespace.LOAD_BALANCERS.ns,
+            Keys.getLoadBalancerKey("*", '*', combineAppStackDetail(applicationName, '*', null))))
+    loadBalancerKeys.addAll(cacheView.filterIdentifiers(Keys.Namespace.LOAD_BALANCERS.ns,
+            Keys.getLoadBalancerKey("*", '*', combineAppStackDetail(applicationName, null, null))))
+
+    cacheView.getAll(Keys.Namespace.LOAD_BALANCERS.ns, loadBalancerKeys)
```
Comment (duftler) [inline clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/view/DcosLoadBalancerProvider.groovy]: Please use `return` statement.

File: clouddriver-dcos/src/test/groovy/com/netflix/spinnaker/clouddriver/dcos/model/DcosLoadBalancerSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+import java.time.Instant
+
+class DcosLoadBalancerSpec extends Specification {
+
+  static final private String APP = "testapp"
+  private static final String ACCOUNT = "testaccount"
+  private static final String REGION = "us-west-1"
+  private static final String LB_NAME = "${APP}-frontend"
+
+  def setup() {
```
Comment (duftler) [inline clouddriver-dcos/src/test/groovy/com/netflix/spinnaker/clouddriver/dcos/model/DcosLoadBalancerSpec.groovy]: Can drop this method.

File: clouddriver-dcos/src/test/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/agent/DcosLoadBalancerCachingAgentSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+    result.evictions == [:]
+  }
+
+  void "On-demand cache should evict a server group that no longer exists"() {
+    given:
+    dcosClient.maybeApp(MARATHON_APP_ID) >> Optional.empty()
+
+    when:
+    def result = subject.handle(providerCache, ["loadBalancerName": LOAD_BALANCER_NAME, "account": ACCOUNT, "region": REGION])
```
Comment (duftler) [inline clouddriver-dcos/src/test/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/agent/DcosLoadBalancerCachingAgentSpec.groovy]: Nice.

File: clouddriver-dcos/src/test/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/agent/DcosLoadBalancerCachingAgentSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+    then:
+    1 * providerCache.evictDeletedItems(Keys.Namespace.ON_DEMAND.ns, [loadBalancerKey])
+    result.evictions == [(Keys.Namespace.LOAD_BALANCERS.ns): [loadBalancerKey]]
+    result.cacheResult.cacheResults.values().flatten() == []
+  }
+
+  void "Newer On-demand cache items should not be replaced by an older version"() {
```
Comment (duftler) [inline clouddriver-dcos/src/test/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/agent/DcosLoadBalancerCachingAgentSpec.groovy]: Nice to see this tested.

**Comment by willgorman** [comment]
Comments have been addressed and commit has been rebased on master.

**Pull request merged**